### PR TITLE
PP-6186 allow the status of a charge to be forcibly updated

### PR DESCRIPTION
**Pull request by oswaldquek**
- Allow a charge to be forcibly transitioned from any state to certain
target states (for now just `CAPTURED`).
- This is to allow for the gateway considering the charge to be in a
different state to what we think it's in.
- This adds new events to be emitted to ledger per status to be
transitioned to. These events need to be added into the enum in the
ledger codebase to link them to the appropriate TransactionStatus.

## WHAT YOU DID
_A brief description of the pull request:_

## How to test

- How should it be reviewed? 
- What feedback do you need? 
- If manual testing is needed, give suggested testing steps.

### Logging

- [ ] only emit log lines at ERROR level which require immediate attention from a support engineer. These will trigger a zendesk alert.

### Documentation

- [ ] Updated README.md for any of the following ?

* Introduced any new environment variables / removed existing environment variable
* Added new API / updated existing API definition

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java [hunk 1]
```diff
@@ -644,6 +648,24 @@ public ChargeEntity transitionChargeState(String chargeId, ChargeStatus targetCh
...
+        ChargeStatus fromChargeState = ChargeStatus.fromString(charge.getStatus());
+
+        return PaymentGatewayStateTransitions.getEventForForceUpdate(targetChargeState).map(eventClass -> {
+            charge.setStatusIgnoringValidTransitions(targetChargeState);
+            ChargeEventEntity chargeEventEntity = chargeEventDao.persistChargeEventOf(charge);
+
+            if (shouldEmitPaymentStateTransitionEvents) {
+                stateTransitionService.offerPaymentStateTransition(
+                        charge.getExternalId(), fromChargeState, targetChargeState, chargeEventEntity,
+                        StatusCorrectedToCapturedToMatchGatewayStatus.class);
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java]: The last argument here should be `eventClass`

**Review by stephencdaly — CHANGES_REQUESTED** [review]
Spotted a small mistake.

Also do you think it's worth adding a test in StateTransitionServiceTest for the new public method we've added?

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java (hunk 1 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java]: Sorry I don't follow... `StatusCorrectedToCapturedToMatchGatewayStatus` is an event class?

**Comment by oswaldquek** [comment]
I don't think we need to add a test. There is a new method but it's just a refactor - pulling out existing code into another which is already tested in https://github.com/alphagov/pay-connector/blob/master/src/test/java/uk/gov/pay/connector/queue/StateTransitionServiceTest.java#L54

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java (hunk 1 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java]: PaymentGatewayStateTransitions.getEventForForceUpdate returns an Optional<Class<T extends Event> which we then map to the variable `eventClass` in the lambda. We should use that variable. The variable will happen to have a value of `StatusCorrectedToCapturedToMatchGatewayStatus.class` always, as currently we're only allowing transitions to `CAPTURED` but the point of the method is we could allow transitions to other states if `getEventForForceUpdate` returns an event class which would be a different class for different target states.

**Comment by stephencdaly** [comment]
> I don't think we need to add a test. There is a new method but it's just a refactor - pulling out existing code into another which is already tested in https://github.com/alphagov/pay-connector/blob/master/src/test/java/uk/gov/pay/connector/queue/StateTransitionServiceTest.java#L54

Yeh that's fair enough

**Review by stephencdaly — APPROVED** [review]

**Pull request merged**